### ISSUE TITLE
Add an option to make Dispose safely dispose of a channel without throwing an exception

### ISSUE THREAD

**Issue by taspeotis** [dotnet/wcf#26]
Using `using` with many WCF objects is a real minefield because `Dispose` can throw exceptions if the channel is in a `Faulted` state. This behavior is well known, but very counter-intuitive and the "pattern" to `try/finally { abort or close }` makes code more verbose than it needs to be.

[By Microsoft's own design guidelines](https://msdn.microsoft.com/en-us/library/bb386039.aspx), `Dispose` should not throw an exception.

In order to avoid breaking [the public contract](https://github.com/dotnet/corefx/wiki/Breaking-Changes#bucket-1-public-contract) (_[a]n exception is no longer thrown_) this behavior could be opt-in.

I do not have a good suggestion for how to provide the opt-in. Setting compatibility settings in `(app|web).config` (like `NetFx40_LegacySecurityPolicy`, `CompatSortNLSVersion`, `TimeSpan_LegacyFormatMode`) will apply to every assembly, when third-party assemblies might be broken by this change (unlikely but possible).

So I submit the following suggestions:
- A `Ca1065CompliantDisposeBehaviorAttribute` on service contracts (channels should know their service contracts?)
- A new constructor/constructor argument/property

**Comment by iamjasonp** [comment]
@taspeotis - yep... we hear you...

This has been a huge gripe for years. For example: [UserVoice](http://aspnet.uservoice.com/forums/147206-wcf-web-services-/suggestions/2635849-fix-the-broken-dispose-pattern-on-communicationobj), [MSDN forums](https://social.msdn.microsoft.com/Forums/vstudio/en-US/99e4f660-23c8-4f7d-bf6b-0ce139aaa2a9/best-way-to-dispose-wcf-service?forum=wcf)... um, people are unhappy.

As you mentioned, breaking public contract is definitely a concern, and that's why the team has hesitated for years on fixing it. Now that we have a chance to start somewhat fresh with .NET Core, it's definitely worth taking another look to see if we can do better. 

Thanks for filing this issue! :+1:

**Comment by mconnew** [comment]
One possibility to achieve what's wanted is to create a custom protocol channel binding element which produces channel factories and channels which wrap the ones from the layer below. See [this blog post](http://blogs.msdn.com/b/carlosfigueira/archive/2011/07/12/wcf-extensibility-channels.aspx) for an example of how to create a custom protocol channel. Typically a protocol binding element has a method similar to this:
```csharp
    public override IChannelFactory<TChannel> BuildChannelFactory<TChannel>(BindingContext context)
    {
        if (typeof(TChannel) != typeof(IRequestChannel))
        {
            throw new InvalidOperationException("Only IRequestChannel is supported");
        }
 
        var factory = new MyFactory(context.BuildInnerChannelFactory<IRequestChannel>());
        return (IChannelFactory<TChannel>)factory;
    }
```
You would implement `MyFactory` such that when `Dispose()` is called on it, it calls `Dispose()` on the inner channel factory wrapped in a try catch. It would be up to the implementation to decide what to do with any exceptions. A similar thing can also be done with the `OnCreateChannel` method in the `MyFactory` class where it would wrap the channel in another channel which passes through all calls except `Dispose()` which would wrap the Try/Catch.  
Basically doing this in a way which pleases everyone in the existing code base would be very difficult. Some people would want to be able to log errors on Dispose but not have it throw. Other people would want only certain exceptions to throw and swallow others.  
We also have the constraint that any changes we make to the API in .NET Core also need to be made in the desktop CLR which would mean aligning any changes with desktop releases.

**Comment by taspeotis** [comment]
> it calls `Dispose()` on the inner channel factory wrapped in a try catch

I know I logged this saying `Dispose` shouldn't throw exceptions (as per the MS guidelines and basically in line with what everybody expects) but I think the following construct avoids throwing any exception when the channel is in the `Faulted` state. That way, really obscure exceptions will still be thrown (although I'm not aware of any other than the `Faulted` ones).

In psuedo-code:

```csharp
if (Channel.State == Faulted)
  Channel.Abort();

Channel.Dispose();
```

That way you save a thrown exception.

**Comment by zhenlan** [comment]
Currently this is worked around in our test code like [this](https://github.com/dotnet/wcf/blob/master/src/System.Private.ServiceModel/tests/Common/Scenarios/ScenarioTestHelpers.cs#L182).

Alternatively, the thinking is to create a custom binding element, which handles any potential exception. This can live in our test code as a good extensibility example.

**Comment by lemoinem** [comment]
Once a strategy to implement `IDisposable` has been agreed upon, in #3676, @mconnew also mentioned the possibility of implementing `IAsyncDisposable`. It could be a nice addition as well, based on `ClosedAsync()`.

Regarding the original discussion:
Couldn't there be a way of having `Dispose()` not throwing (potentially silently discarding any exception), while giving a way to client code to check for and trigger these exceptions if/when wanted?


For example, based on https://github.com/dotnet/wcf/issues/26#issuecomment-127425033:

The base/default channel would have a `TryDispose()` method, that's equivalent to the current `Dispose()` (i.e. can throw if/when an issue arises).
Its `Dispose()` would be calling `TryDispose()` and silently discarding any exception, ensuring any resources are still freed at the time (thus completely fitting the `IDisposable` contract).

If a client needs a different behavior, then they would have to implement a custom channel and in their `Dispose()`, call `TryDispose()` and deal with any exception as they see fit.


Another solution: Provide a "OnExceptionWhileDisposingHandler" Property on the channel (client?). When Dispose would throw, it will call the delegate, let the client code handle the exception in any way (default handler just drops the exception).
The handler could also return true/false depending on whether it considers the error irrecoverable (in which case, `Abort()` should be called by `Dispose()`) or if it was able to recover (in which case `Close()` could be called again to try to close the channel gracefully). I'm not sure if this last part is actually useful.


What do you think?

**Comment by drdamour** [comment]
what about just starting with an extension method that does it the best practice way?
